org.adempiere.base/src/org/compiere/model/MProductionLine.java

```diff
@@ -1,457 +1,519 @@
+			productionParent = new MProduction(getCtx(), getM_Production_ID(), get_TrxName());
+
+		if (getM_Production_ID() > 0) 
+		{
+			if ( productionParent.getM_Product_ID() == getM_Product_ID() && productionParent.getProductionQty().signum() == getMovementQty().signum())
+				setIsEndProduct(true);
+			else 
+				setIsEndProduct(false);
+		} 
+		else 
+		{
+			I_M_ProductionPlan plan = getM_ProductionPlan();
+			if (plan.getM_Product_ID() == getM_Product_ID() && plan.getProductionQty().signum() == getMovementQty().signum())
+				setIsEndProduct(true);
+			else 
+				setIsEndProduct(false);
+		}
+		
+		if ( isEndProduct() && getM_AttributeSetInstance_ID() != 0 )
+		{
+			String where = "M_QualityTest_ID IN (SELECT M_QualityTest_ID " +
+			"FROM M_Product_QualityTest WHERE M_Product_ID=?) " +
+			"AND M_QualityTest_ID NOT IN (SELECT M_QualityTest_ID " +
+			"FROM M_QualityTestResult WHERE M_AttributeSetInstance_ID=?)";
+
+			List<MQualityTest> tests = new Query(getCtx(), MQualityTest.Table_Name, where, get_TrxName())
+			.setOnlyActiveRecords(true).setParameters(getM_Product_ID(), getM_AttributeSetInstance_ID()).list();
+			// create quality control results
+			for (MQualityTest test : tests)
+			{
+				test.createResult(getM_AttributeSetInstance_ID());
+			}
+		}
+		
+		if ( !isEndProduct() )
+		{
+			setMovementQty(getQtyUsed().negate());
+		}
+		
+		return true;
+	}
+	
+	@Override
+	protected boolean beforeDelete() {
+		
+		deleteMA();
+		return true;
+	}
+
+	/**
+	 * 	Get Reversal_ID of parent production
+	 *	@return Reversal_ID
+	 */
+	public int getProductionReversalId() {
+		if (getM_Production_ID() > 0)
+			return DB.getSQLValueEx(get_TrxName(), "SELECT Reversal_ID FROM M_Production WHERE M_Production_ID=?", getM_Production_ID());
+		else
+			return DB.getSQLValueEx(get_TrxName(), "SELECT p.Reversal_ID FROM M_ProductionPlan pp INNER JOIN M_Production p ON (pp.M_Production_ID = p.M_Production_ID) WHERE pp.M_ProductionPlan_ID=?", getM_ProductionPlan_ID());
+	}
+
+	/**
+	 * 
+	 * @return
+	 */
+	public MProductionLineMA[] getLineMAs() {
+		ArrayList<MProductionLineMA> list = new ArrayList<MProductionLineMA>();
+		
+		String sql = "SELECT pl.M_ProductionLine_ID, pl,M_AttributeSetInstance_ID , pl.MovementQty, pl.DateMaterialPolicy "
+			+ "FROM M_ProductionLineMA pl "
+			+ "WHERE pl.M_ProductionLine_ID = ?";
+		
+		PreparedStatement pstmt = null;
+		ResultSet rs = null;
+		try
+		{
+			pstmt = DB.prepareStatement(sql, get_TrxName());
+			pstmt.setInt(1, get_ID());			
+			rs = pstmt.executeQuery();
+			while (rs.next())
+				list.add( new MProductionLineMA( this, rs.getInt(2), rs.getBigDecimal(3), rs.getTimestamp(4) ) );	
+		}
+		catch (SQLException ex)
+		{
+			throw new AdempiereException("Unable to load production lines", ex);
+		}
+		finally
+		{
+			DB.close(rs, pstmt);
+			rs = null;
+			pstmt = null;
+		}
+		
+		MProductionLineMA[] retValue = new MProductionLineMA[list.size()];
+		list.toArray(retValue);
+		return retValue;
+	}
+}
```

Comment:
I'm sure you didn't change the whole file. We can't review when changes/PR is not submitted as a proper diff.